delivery_sendcloud_oca/models/delivery_carrier.py

```diff
@@ -0,0 +1,476 @@
+    # -------------------------- #
+
+    def sendcloud_rate_shipment(self, order):
+        self.ensure_one()
+        res = {
+            "success": False,
+            "price": 0.0,
+            "warning_message": False,
+            "error_message": False,
+        }
+        if not order.partner_shipping_id.country_id:
+            res["error_message"] = _("Partner does not have any country.")
+            return res
+
+        country = order.partner_shipping_id.country_id
+        price, method_country = self._sendcloud_get_price_per_country(country.code)
+        price_digits = self.env["decimal.precision"].precision_get("Product Price")
+        price = float_round(price, precision_digits=price_digits)
+        price = order._sendcloud_convert_price_in_euro(price)
+        res["success"] = True
+        res["price"] = price
+        res["sendcloud_country_specific_product"] = method_country.product_id
+
+        if self.sendcloud_service_point_input == "required":
+            res["warning_message"] = _("This shipping method requires a Service Point.")
+
+        return res
+
+    def sendcloud_send_shipping(self, pickings):
+        self.ensure_one()
+        # Dispatch Order and create labels
+        res = pickings._sendcloud_send_shipping()
+        parcels = pickings.mapped("sendcloud_parcel_ids")
+        parcels._generate_parcel_labels()
+        return res
+
+    def sendcloud_cancel_shipment(self, picking):
+        ctx = {"skip_sync_picking_to_sendcloud": True, "skip_cancel_parcel": True}
+        deleted_parcels = []
+        for parcel_code in picking.mapped("sendcloud_parcel_ids.sendcloud_code"):
+            integration = picking.company_id.sendcloud_default_integration_id
+            res = integration.cancel_parcel(parcel_code)
+            if res.get("status") == "deleted":
+                deleted_parcels.append(parcel_code)
+            elif (
+                res.get("status") == "failed"
+                and res.get("message") == "This shipment is already being cancelled."
+            ):
+                deleted_parcels.append(parcel_code)
+            elif res.get("error", {}).get("code") == 404:
+                deleted_parcels.append(parcel_code)  # ignore "Not Found" error
+            elif res.get("error"):
+                raise ValidationError(_("Sendcloud: %s") % res["error"].get("message"))
+        parcels_to_delete = picking.sendcloud_parcel_ids.filtered(
+            lambda p: p.sendcloud_code in deleted_parcels
+        )
+        parcels_to_delete.with_context(**ctx).unlink()
+        picking.with_context(**ctx).write({"carrier_price": 0.0})
+
+    # ------------------------------- #
+    # Inherits for Sendcloud provider #
+    # ------------------------------- #
+
+    def _compute_can_generate_return(self):
+        res = super()._compute_can_generate_return()
+        for carrier in self.filtered(lambda c: c.delivery_type == "sendcloud"):
+            carrier.can_generate_return = True
```

Comment:
you can use an update() on recordset instead.

Reply:
Done !!